pennylane/transforms/batch_input.py

```diff
@@ -0,0 +1,87 @@
+"""
+    Batch transformation for multiple (non-trainable) input examples following issue #2037
+"""
+from typing import Union, Sequence, Callable, Tuple
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.transforms.batch_transform import batch_transform
+
+
+@batch_transform
+def batch_input(
+    tape: Union[qml.tape.JacobianTape, qml.QNode],
+    argnum: Union[Sequence[int], int] = 0,
+) -> Tuple[Sequence[qml.tape.JacobianTape], Callable]:
+    """
+    In a classical ML application one needs to batch the non-trainable inputs of the network.
+    This function executes the same analogue for a quantum circuit where separate circuit
+    executions are created for each input examples  which will be executed with the same
+    trainable inputs. The batch dimension assummed to be the first rank of the
+    non trainable tensor object where for a rank 1 feature space the shape needs to be (Nt, x)
+    where `x` indicates the dimension of the features and `Nt` being the number of examples
+    within a batch.
+    Based on https://arxiv.org/abs/2202.10471
+
+    * Example *
+
+    .. code-block:: python
+
+            dev = qml.device("default.qubit", wires = 2, shots=None)
+            @batch_input(argnum=0)
+            @qml.qnode(dev, diff_method="parameter-shift", interface="tf")
+            def circuit(inputs, weights):
+                qml.AngleEmbedding(inputs, wires = range(2), rotation="Y")
+                qml.RY(weights[0], wires=0)
+                qml.RY(weights[1], wires=1)
+                return qml.expval(qml.PauliZ(1))
+
+            >>> x = np.random.uniform(0,1,(10,2))
+            >>> x.requires_grad = False
+            >>> w = np.random.uniform(0,1,2)
+            >>> circuit(x, w)
+            <tf.Tensor: shape=(10,), dtype=float64, numpy=
+            array([0.17926078, 0.7480163 , 0.47816999, 0.50381628, 0.349178  ,
+                   0.17511444, 0.03769436, 0.19180259, 0.75867188, 0.55335748])>
+
+
+    Parameters
+    ----------
+    tape (qml.tape.JacobianTape or qml.QNode): Record of the inputs to be executed
+    argnum (Sequence[int] or int) : One or more argument numbers indicating the
+        location of the batched inputs. As default first argument is assumed to be the only
+        batched input.
+
+    Returns
+    -------
+    Sequence[Sequence[qml.tape.JacobianTape], Callable]
+        list of tapes arranged according to unbatched inputs and a callable function
+        to batch the results.
+    """
+    parameters = tape.get_parameters(trainable_only=False)
+
+    argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
+
+    non_trainable, trainable = [], []
```

Comment:
technically, these would `non_batched` and `batched`, right? Since it is possible for a user to use this function to batch a trainable parameter, from my read of it

Reply:
No, I believe argnum should be used only for non-trainable inputs, [please see modified comment here](https://github.com/PennyLaneAI/pennylane/pull/2069#discussion_r818583025). I believe this is more descriptive. To use with trainable parameters you already have `batch_params` function so I believe having a clear distinction is better. what do you think?

Reply:
Ah, this is likely because of how `tape.get_parameters()` works.

- With `trainable_only=True`, it returns only the trainable parameters.
- With `trainable_only=False`, it returns _all_ parameters on the tape, both trainable and non-trainable.

Since `batch_input` uses the latter, `argnum` will be the index value on _all_ gate parameters, independent if it is trainable or not.

So it might be best to indicate this in the documentation.

For example,

```python
@qml.batch_inputs(argnum=1)
def circuit():
    # index 0
    qml.RX(np.array(0.1, requires_grad=False), wires=0)

    # index 1, will apply to this gate argument even if trainable.
    qml.RX(np.array([0.1, 0.2], requires_grad=True), wires=0)
```

Reply:
> Ah, this is likely because of how `tape.get_parameters()` works.
> 
> * With `trainable_only=True`, it returns only the trainable parameters.
> * With `trainable_only=False`, it returns _all_ parameters on the tape, both trainable and non-trainable.

Exactly, would have been easier if there was something like the `non_trainable_only` boolean option. Do you think it would be good to add such an option to `tape`?

> Since `batch_input` uses the latter, `argnum` will be the index value on _all_ gate parameters, independent if it is trainable or not.
> 
> So it might be best to indicate this in the documentation.
> 
> For example,
> 
> ```python
> @qml.batch_inputs(argnum=1)
> def circuit():
>     # index 0
>     qml.RX(np.array(0.1, requires_grad=False), wires=0)
> 
>     # index 1, will apply to this gate argument even if trainable.
>     qml.RX(np.array([0.1, 0.2], requires_grad=True), wires=0)
> ```
Aha, this is great input thanks a lot I totally missed this possibility!
definition modified accordingly:
```
        argnum (Sequence[int] or int): One or more index value on all gate parameters
            indicating the location of the non-trainable batched inputs within the input
            argument sequence of the circuit. By default first argument is assumed to be
            the only batched input.
```

Reply:
Only modified the docstring for this, let me know if you think expanding the options for `tape.get_parameters()` is a better way to go.